ui/app/adapters/job.js

```diff
@@ -159,4 +167,138 @@ export default class JobAdapter extends WatchableNamespaceIDs {
       },
     });
   }
+
+  runAction(job, action, allocID) {
+    let messageBuffer = '';
+
+    const protocol = window.location.protocol === 'https:' ? 'wss:' : 'ws:';
+    const region = this.system.activeRegion;
+    const applicationAdapter = getOwner(this).lookup('adapter:application');
+    const prefix = `${
+      applicationAdapter.host || window.location.host
+    }/${applicationAdapter.urlPrefix()}`;
+
+    const wsUrl =
+      `${protocol}//${prefix}/job/${encodeURIComponent(job.get('id'))}/action` +
+      `?namespace=${job.get('namespace.id')}&action=${
+        action.name
+      }&allocID=${allocID}&task=${action.task.name}&group=${
+        action.task.taskGroup.name
+      }&tty=true&ws_handshake=true` +
+      (region ? `&region=${region}` : '');
+
+    // const socket = new WebSocket(wsUrl);
+    let socket;
+
+    const mirageEnabled =
+      config.environment !== 'production' &&
+      config['ember-cli-mirage'] &&
+      config['ember-cli-mirage'].enabled !== false;
+
+    if (mirageEnabled) {
+      socket = new Object({
+        messageDisplayed: false,
+        addEventListener: function (event, callback) {
+          if (event === 'message') {
+            this.onmessage = callback;
+          }
+          if (event === 'open') {
+            this.onopen = callback;
+          }
+          if (event === 'close') {
+            this.onclose = callback;
+          }
+        },
+
+        send(e) {
+          if (!this.messageDisplayed) {
+            this.messageDisplayed = true;
+            this.onmessage({
+              data: `{"stdout":{"data":"${btoa('Message Received')}"}}`,
+            });
+          } else {
+            this.onmessage({ data: e.replace('stdin', 'stdout') });
+          }
+        },
+      });
+    } else {
+      socket = new WebSocket(wsUrl);
+    }
+
+    let notification;
+    socket.addEventListener('open', () => {
+      notification = this.notifications
+        .add({
+          title: `Action ${action.name} Started`,
+          color: 'neutral',
+          code: true,
+          sticky: true,
+          customAction: {
+            label: 'Stop Action',
+            action: () => {
+              socket.close();
+            },
+          },
+        })
+        .getFlashObject();
+
+      socket.send(
+        JSON.stringify({ version: 1, auth_token: this.token?.secret || '' })
+      );
+      socket.send(
+        JSON.stringify({
+          tty_size: { width: 250, height: 100 }, // Magic numbers, but they pass the eye test.
+        })
```

Comment:
I believe this is what we use the xterm-addon-fit for over in [ExecTerminal](https://github.com/hashicorp/nomad/blob/1ef99f05364b7d3739befa6a789f0d55b2314dcf/ui/app/components/exec-terminal.js).